qa/rpc-tests/p2p-addr.py

```diff
@@ -0,0 +1,168 @@
+
+    def last_port_sent(self):
+        assert self.counter > 0
+        return self.index_to_port(self.counter - 1)
+
+    def have_received_port(self, port):
+        for peer in self.recv_nodes:
+            if port in peer.ports_received:
+                return True
+        return False
+
+    def wait_for_specific_port(self, port):
+        def port_received():
+            return self.have_received_port(port)
+        return wait_until(port_received, timeout=600)
+
+    def create_addr_msg(self, num, services):
+        addrs = []
+        for i in range(num):
+            addr = CAddress()
+            addr.time = self.mocktime + random.randrange(-100, 100)
+            addr.nServices = services
+            assert self.counter < 256 ** 2  # Don't allow the returned ip addresses to wrap.
+            addr.ip = f"123.123.{self.counter // 256}.{self.counter % 256}"
+            addr.port = self.index_to_port(self.counter)
+            self.counter += 1
+            addrs.append(addr)
+
+        msg = msg_addr()
+        msg.addrs = addrs
+        return msg
+
+    def send_addr_msg(self, msg):
+        self.send_node.connection.send_message(msg)
+        time.sleep(0.5) # sleep half a second to prevent mocktime racing the msg
+
+        # invoke m_next_addr_send timer:
+        # `addr` messages are sent on an exponential distribution with mean interval of 30s.
+        # Setting the mocktime 600s forward gives a probability of (1 - e^-(600/30)) that
+        # the event will occur (i.e. this fails once in ~500 million repeats).
+        self.mocktime += 60 * 10
+        self.nodes[0].setmocktime(self.mocktime)
+
+        time.sleep(0.5) # sleep half a second to prevent pings racing mocktime
+        for peer in self.recv_nodes:
+            peer.sync_with_ping()
+
+    def create_and_send_addr_msg(self, num, services=NODE_NETWORK):
+        self.send_addr_msg(self.create_addr_msg(num, services))
+
+    def simple_relay_test(self):
+        # send a message with 2 addresses
+        self.create_and_send_addr_msg(2)
+
+        # make sure we received the last addr record
+        assert self.wait_for_specific_port(self.last_port_sent())
+
+    def oversized_addr_test(self):
+        # create message with 1010 entries and
+        # confirm that the node discarded the entries
+
+        # send one valid message, keep track of the port it contains
+        valid_port_before = self.index_to_port(self.counter)
+        self.create_and_send_addr_msg(1)
+
+        # send a too large message that will be ignored
+        self.create_and_send_addr_msg(1010)
+
+        # finish with a valid message, keep track of the port it contains
+        valid_port_after = self.index_to_port(self.counter)
+        self.create_and_send_addr_msg(1)
+
+        # wait until both valid addresses were propagated
+        assert self.wait_for_specific_port(valid_port_before)
+        assert self.wait_for_specific_port(valid_port_after)
```

Comment:
This isn't a blocker for me.

I understand and agree with the goal here, though I wonder if rewriting `wait_for_specific_port` to wait for _multiple_ ports would be in our favor here. I don't know that there's any debugging value in knowing one or the other address was or wasn't propagated, and repeating the potential 10 minute wait could be expensive.

It's probably not an issue, but if we ever see any issues with this, I want to leave this comment as a reminder of this idea.